newIDE/app/src/Leaderboard/LeaderboardProvider.js

```diff
@@ -0,0 +1,406 @@
+      };
+    case 'UPDATE_OR_CREATE_LEADERBOARD':
+      return {
+        ...state,
+        displayOnlyBestEntry: shouldDisplayOnlyBestEntries(action.payload),
+        leaderboardsByIds: {
+          ...state.leaderboardsByIds,
+          [action.payload.id]: action.payload,
+        },
+        currentLeaderboardId: action.payload.id,
+        currentLeaderboard: action.payload,
+      };
+    case 'REMOVE_LEADERBOARD':
+      const newLeaderboardsByIds = { ...state.leaderboardsByIds };
+      delete newLeaderboardsByIds[action.payload];
+      const leaderboardsIds = Object.keys(newLeaderboardsByIds);
+      if (leaderboardsIds.length === 0) {
+        return {
+          ...state,
+          displayOnlyBestEntry: false,
+          leaderboardsByIds: newLeaderboardsByIds,
+          currentLeaderboard: null,
+          currentLeaderboardId: null,
+        };
+      }
+      return {
+        ...state,
+        displayOnlyBestEntry: shouldDisplayOnlyBestEntries(
+          newLeaderboardsByIds[leaderboardsIds[0]]
+        ),
+        leaderboardsByIds: newLeaderboardsByIds,
+        currentLeaderboard: newLeaderboardsByIds[leaderboardsIds[0]],
+        currentLeaderboardId: leaderboardsIds[0],
+      };
+    default: {
+      throw new Error(`Unhandled type: ${action.type}`);
+    }
+  }
+};
+
+const LeaderboardProvider = ({ gameId, children }: Props) => {
+  const authenticatedUser = React.useContext(AuthenticatedUserContext);
+
+  const [
+    {
+      currentLeaderboardId,
+      currentLeaderboard,
+      leaderboardsByIds,
+      displayOnlyBestEntry,
+      entries,
+      mapPageIndexToUri,
+      pageIndex,
+    },
+    dispatch,
+  ] = React.useReducer<ReducerState, ReducerAction>(reducer, {
+    currentLeaderboardId: null,
+    currentLeaderboard: null,
+    leaderboardsByIds: null,
+    displayOnlyBestEntry: false,
+    entries: null,
+    mapPageIndexToUri: {},
+    pageIndex: 0,
+  });
+
+  const listLeaderboards = React.useCallback(
+    async () => {
+      dispatch({ type: 'SET_LEADERBOARDS', payload: null });
+      const fetchedLeaderboards = await listGameLeaderboards(gameId);
+      fetchedLeaderboards.sort((a, b) => a.name.localeCompare(b.name));
+      dispatch({ type: 'SET_LEADERBOARDS', payload: fetchedLeaderboards });
+    },
+    [gameId]
+  );
+
+  const createLeaderboard = React.useCallback(
+    async (creationPayload: {|
+      name: string,
+      sort: LeaderboardSortOption,
+    |}) => {
+      const newLeaderboard = await doCreateLeaderboard(
+        authenticatedUser,
+        gameId,
+        creationPayload
+      );
+      if (!newLeaderboard) return;
+
+      dispatch({
+        type: 'UPDATE_OR_CREATE_LEADERBOARD',
+        payload: newLeaderboard,
+      });
+    },
+    [gameId, authenticatedUser]
+  );
+
+  const fetchEntries = React.useCallback(
+    async (options?: {| uri?: ?string |}) => {
+      if (!currentLeaderboardId) return;
+
+      let uriToUse = null;
```

Comment:
you don't need a let and can probably do `const utiToUse = options && options.uri ? options.uri || null` or something like that